ioBroker/ioBroker.node-red
# Add HTTPS support to ioBroker.Node-RED based on ioBroker Certificates

Hi @GermanBluefox , @Apollon77,
as mentioned in PR #99 and as initially mentioned in my feature request #97 I created another PR to add HTTPS support. Important remark: I forgot to branch PR #99 and implemented this PR afterwards. Hence, this PR contains the changes of #99 as well and should be checked and merged after merging #99. To see the real changes of this PR, please check: https://github.com/jwiesel/ioBroker.node-red/compare/master...jwiesel:https (Commits since April 14)

## Implemented changes
* Added https specific fields to settings page index_m.html - code has been taken from iobroker.web adapter
* Added fields for https, certPublic and certPrivate to iopackage.json
* updated writeSettings function in main.js to load the selected certificates (if selected) and store them in the iobroker data folder. --> directly injecting the certificate and key into the settings.js didn't really work due to multi-line string issues
* updated file settings.js to inject certificate and key path and uncomment `var fs` and section `https:`, if HTTPS is enabled in settings.

## Testing done so far
* I did various tests by installing the adapter into ioBroker from my Github fork (`npm install https://github.com/jwiesel/ioBroker.node-red/tarball/https`) as indicated in the ioBroker documentation: https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/adapterdev.md#deploying
 
## Test Instructions
* Install the adapter in your ioBroker installation
* Activate HTTPS and select certificate and key in Node-RED instance settings. Check the log to see, that Node-RED is starting up without errors
* Open node-red: Caution: The URL will now start with https:// instead of http:// --> there is also no redirect possible as one port can only support one protocol --> now node-red should load
* Close Node-RED again
* Go to the instance settings and remove certificate and key + disable HTTPS and save the settings. Wait a few seconds until Node-RED is up and running again and check the log for errors (there should be any)
* Open node-red now via the http:// address again, as https:// is disabled --> Node-Red should run via HTTP again

## Testing the node-red link in iobroker admin sidebar
* The admin sidebar of iobroker is not updated directly after changing instance settings.
* To refresh the link to Node-Red in the admin menu after enabling HTTPS the entire admin page has to be refreshed or the user can logout and login again to see the effect of the change.
* As most users will just enable https once, this should not be an issue
